File: 3.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strings"
)

func readInput(path string) ([][]string, error) {
  file, err := os.Open(path)

  if err != nil {
    return nil, err
  }
  defer file.Close()

  var lines [][]string
  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    lines = append(lines, strings.Split(scanner.Text(), ""))
  }

  return lines, scanner.Err()
}

func partOneSolution(sample [][]string) (int) {
  var treeCount int = 0
  startIdx := 0
  step := 3

  for i := 0; i < len(sample) - 1; i++ {
    row := sample[i]
    startIdx = int(math.Mod(float64(startIdx + step), float64(len(row))));

    if sample[i + 1][startIdx] == "#" {
      treeCount++
    }
  }
  return treeCount
}

func main() {
    sample, e  := readInput("input-3.txt")
    if e != nil {
      fmt.Println("Reading input", e)
    }
    treeCount := partOneSolution(sample)
    fmt.Printf("tree count: %d\n", treeCount)
}

File: 3.hs
len :: [a] -> Int
len = len' 0
  where
    len' a [] = a
    len' a (_ : xs) = len' (a + 1) xs

getTreeCount :: (Show t, Num t) => [[Char]] -> Int -> Int -> Int -> Int -> Int -> t -> IO ()
getTreeCount sample index size step down startIdx treeCount = do
  if (index + 1) >= size
    then do
      putStrLn ("slope (step: " ++ show step ++ ", down: " ++ show down ++ ") " ++ "tree count :>> " ++ show treeCount)
    else do
      if sample !! (index + down) !! (startIdx) == '#'
        then do
          let updatedStartIdx = (index + step) `mod` len (sample !! index)
          getTreeCount sample (index + down) size step down updatedStartIdx (treeCount + 1)
        else do
          getTreeCount sample (index + down) size step down startIdx treeCount

getTotalTreesTraversedBySlope :: [[Char]] -> Int -> Int -> IO ()
getTotalTreesTraversedBySlope sample step down = do
  let size = len sample
  getTreeCount sample 0 size step down 0 0

main :: IO ()
main = do
  let sample = ["#ab", "#ab", "abc"]
  getTotalTreesTraversedBySlope sample 1 1
  getTotalTreesTraversedBySlope sample 3 1
  getTotalTreesTraversedBySlope sample 5 1
  getTotalTreesTraversedBySlope sample 7 1
  getTotalTreesTraversedBySlope sample 1 2
